### ISSUE TITLE
Objective functions

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#15]
There is currently a submodule `nntools.objectives`, but like the regularization one this was a bit of an afterthought.

I would be great if the code in `nntools.objectives` is defined in terms of Theano expressions (not `Layer` instances), so that it can be used in isolation.

But sometimes it might be convenient to have an object that represents the objective function, which defines a Theano variable for the target labels (in the case of a supervised training objective), so that you can just do something like:
```
l_out = ... # output layer of the network
obj = nntools.objectives.Objective(l_out, loss_function=nntools.objectives.mse)
loss = obj.get_loss()
```
I don't know if this warrants the extra class though. It doesn't really do that much. What does everyone think?

**Comment by ebattenberg** [comment]
I wanted to revive this discussion with something I came across.  In my work, I sometimes use a weighted loss function where one might weight individual examples differently or weight individual output labels or classes differently.  For this, I would need to add additional arguments to the loss function besides `x` and `t` (the predicted and target outputs).  

So it would be nice to be able to pass additional parameters to the loss function I define.  Right now, the `Objective` constructor passes all additional args to `get_output`.  

I guess I could do this using a function closure as well.  Or I could implement it separately since the `Objective` class is more of a convenience function.

**Comment by benanne** [comment]
Have a look at the discussion on #127 :)

**Comment by benanne** [comment]
The `lasagne.objectives` module has been slightly reworked in the meantime so that the various objective functions (mse etc.) no longer compute the average across the batch. This is taken care of by the `Objective` class instead, which has allowed us to implement a `MaskedObjective`.

The question was raised whether we really need `Objective` objects - they only really define one method after all (`get_loss()`). Maybe their job could be done by simple Python functions instead.

The reason I like `Objective` as a class is because it allows for a default target variable (`target_var`) to be defined automatically, so I don't need to worry about this. This is nicely symmetric with the input variable (`input_var`) that is automatically created by the `InputLayer`. So I think I'm in favour of keeping things like they are for this reason. But we agreed during the dev meeting that this should be discussed, so: discuss :)

**Comment by f0k** [comment]
> The question was raised whether we really need `Objective` objects - they only really define one method after all (`get_loss()`). Maybe their job could be done by simple Python functions instead.

For me the current `Objective` or `MaskedObjective` classes are not really helpful, as they do too much at once: Obtaining the output of a given layer with a given input expression, calling a single loss function with that and an automatically-created or given target, and then averaging, summing or weighted-averaging the item-wise losses.

I like that we have loss functions that take Theano expressions and return the item-wise losses, so I can freely combine them in whatever way I find useful. It would be nice to have similarly generic functions that just apply the (normal, masked or weighted) accumulation. Granted, there wouldn't be too much of a benefit over just writing `(myloss * mymask).mean()` directly.

Anyway, for the simple use case of having a single-output network and a single loss on that (rather than a combination of those), these classes might be nice to have, and I understand your "symmetry" argument about creating the target variable automatically (which requires them to be classes). What I don't like is that so many arguments go both into the constructor and into `get_loss()`, with the latter overwriting choices of the former if given. I think that everything should be in constructor arguments, reducing `get_loss()` to `get_loss(self)` for all objective classes. Furthermore, I'd suggest to rename `input_layer` to `network` or `network_output_layer`, to make it less confusing.

> discuss :)

Duty fulfilled :)

**Comment by benanne** [comment]
> For me the current Objective or MaskedObjective classes are not really helpful, as they do too much at once: Obtaining the output of a given layer with a given input expression, calling a single loss function with that and an automatically-created or given target, and then averaging, summing or weighted-averaging the item-wise losses.

Fair point, it seems like some of these steps may be useful on their own and could be isolated if we come up with a better API.

> It would be nice to have similarly generic functions that just apply the (normal, masked or weighted) accumulation.

I agree. Can we do that easily without making the typical use cases too complex? 

>  What I don't like is that so many arguments go both into the constructor and into get_loss(), with the latter overwriting choices of the former if given.

Agree, that seems a bit redundant. The little bit of additional flexibility this offers is pretty detrimental to code quality and readability in this case. Renaming `input_layer` is also a good idea since this does indeed confuse people (see #173).

I'll see if I can draft a PR later this week to address some of this stuff.

**Comment by benanne** [comment]
The objectives submodule has been revamped in #286.
